### ISSUE TITLE
COPY on append partitioned table whose partition column is composite type crashes

### ISSUE THREAD

**Issue by byucesoy** [citusdata/citus#728]
Server crashes when we try to COPY to append partitioned table whose partition column is composite type **if** min/max functions is not defined for that type. Instead of crashing, we should error out.

To reproduce the bug

```sql
--run at master & workers, citus does not propogate CREATE TYPE yet
CREATE TYPE new_composite_type as (key1 text, key2 text);

CREATE TABLE nation_append (
    n_nationkey integer not null,
    n_name char(25) not null,
    n_regionkey integer not null,
    n_comment varchar(152),
    test_col new_composite_type
 );

SELECT master_create_distributed_table('nation_append', 'test_col', 'append');

\COPY nation_append FROM STDIN WITH CSV
1,'name',1,'comment_1',"(a,a)"
2,'name',2,'comment_2',"(a,b)"
3,'name',3,'comment_3',"(a,c)"
4,'name',4,'comment_4',"(a,d)"
5,'name',5,'comment_5',"(a,e)"

-- master output
WARNING:  function min(new_composite_type) does not exist
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
CONTEXT:  while executing command on ip-10-111-166-135.ec2.internal:5432
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.

-- worker output
ERROR:  function min(new_composite_type) does not exist at character 8
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
STATEMENT:  SELECT min(test_col), max(test_col) FROM public.nation_append_103298
```

This bug does not exist in 5.1 release. In 5.1 we give this error message;
```sql
WARNING:  could not receive query results from ip-10-218-175-197.ec2.internal:5432
DETAIL:  Client error: function min(new_composite_type) does not exist
WARNING:  could not receive query results from ip-10-136-52-162.ec2.internal:5432
DETAIL:  Client error: function min(new_composite_type) does not exist
WARNING:  could not get statistics for shard public.nation_append_102110
DETAIL:  Setting shard statistics to NULL
COPY 5
```

**Comment by ozgune** [comment]
@byucesoy / @metdos -- how involved is the fix for this issue?

**Comment by metdos** [comment]
@ozgune, @sumedhpathak 

The root cause of this bug resides MultiClientQueryResult(). Basically, we should return false if the query fails. However, we return true at the time being.

I created the PR #732 to fix this which just adds a line to return immediately false if the query fails.

This is a simple fix but effects multiple code paths. Should we include this fix in 5.2?

**Comment by metdos** [comment]
Note that this problem does not exist for hash-partitioned tables, because we don't call WorkerShardStats() on hash-partitioned tables.

**Comment by ozgune** [comment]
@metdos -- thanks for the update! From a timeline standpoint, it would be nice (but not a must) to fix this before our v5.2 release.

Could you elaborate a bit more on the following?

* Why did we start observing this issue in v5.2? If this functionality was working in v5.1, what change triggered this bug?

* If I recall correctly, when we originally introduced ``MultiClientQueryResult()``, we assumed that we'd only need to read results from ``SELECT`` statements in this function. We therefore [only checked against](https://www.postgresql.org/docs/current/static/libpq-exec.html) the ``PGRES_TUPLES_OK`` case. The additional else statement in ``MultiClientQueryResult()`` was only for defensive coding. Do you know if that assumption changed?

**Comment by metdos** [comment]
> Why did we start observing this issue in v5.2? If this functionality was working in v5.1, what change triggered this bug?

Program crashes in WorkerShardStats() while trying to read/free query result on a failing query. 

The code path is changed in PR #670.

We used to call  WorkerShardStats() -> WorkerPartitionValue() -> ExecuteRemoteQuery()
Now we call        WorkerShardStats()-> MultiClientExecute() -> MultiClientQueryResult()

> If I recall correctly, when we originally introduced MultiClientQueryResult(), we assumed that we'd only need to read results from SELECT statements in this function. We therefore only checked against the PGRES_TUPLES_OK case. The additional else statement in MultiClientQueryResult() was only for defensive coding. Do you know if that assumption changed?

Should not we return false if PGRES_TUPLES_OK is not the case? Currently, MultiClientQueryResult() returns true even though PGRES_TUPLES_OK is not the case. Was this the original assumption?

**Comment by marcocitus** [comment]
I think MultiClientQueryResult qualifies for the underhanded C competition :). Returning false seems like the right fix.

We'll probably remove the MutliClient* API as part of connection management unification since it's no longer necessary for calling libpq and it's a source of bugs.

**Comment by ozgune** [comment]
>> Should not we return false if PGRES_TUPLES_OK is not the case? Currently, MultiClientQueryResult() returns true even though PGRES_TUPLES_OK is not the case. Was this the original assumption?

I originally assumed that we'd initiate a ``SELECT``, read query results through ``MultiClientQueryResult()``, and therefore would never fall into the ``else`` block. Clearly, that assumption isn't valid anymore. :)

@metdos / @marcocitus -- sounds good on the fix. I'll just add a note to the pull request.

**Comment by ozgune** [comment]
We chatted about this issue's impact here and decided that we didn't need to fix before our v5.2 release. I'm adding the 5.3 milestone to this issue.
